### PR TITLE
ENH: Add built-in function for Styler to format the text displayed for missing values

### LINKED ISSUE #21527: ENH: Styler display of NaN / null Values
The `Styler` allows for null cells to be highlighted using `.style.highlight_null()` but a `nan` text string will also be displayed.

If a column is coloured using `.background_gradient()` it must be a numeric type otherwise numerical comparisons fail. This means that using a construction of the form `..fillna('')` on the column will create an error as the column is cast to an `object` type.

`nan` values can be removed in the rendered table using a hack of the form `.render().replace('nan','')` but it would be cleaner to use something along the lines of `.style.replace_null(value)` that could be used to tidy up the display of `nan` cells.

### LINKED ISSUE #28358: passing string value to pandas.DataFrame.fillna()   &   pandas.PivotTable(fill_value) 'breaks' pandas.DataFrame.style.highlight_* inside jupyter notebook
#### Code Sample, a copy-pastable example if possible

```python
import pandas as pd
import numpy as np
df = pd.DataFrame({"A": ["foo", "foo", "foo", "foo", "foo",
                          "bar", "bar", "bar", "bar"],
                    "B": ["one", "one", "one", "two", "two",
                          "one", "one", "two", "two"],
                    "C": ["small", "large", "large", "small",
                          "small", "large", "small", "small",
                          "large"],
                    "D": [1, 2, 2, 3, 3, 4, None, 6, 7],
                    "E": [2, 4, 5, 5, 6, 6, 8, 9, 9]})
table = pd.pivot_table(df, values='D', index=['A','B'], columns=['C'], aggfunc=np.sum, fill_value='-')

df.fillna('-').style.highlight_max()
table.style.hightlight_max()


```
Screenshot from 2019-09-09 11-07-48

Screenshot from 2019-09-09 11-07-23

#### Problem description
`pandas.DataFrame.style.highlight_*` does not work on column where `nan` has been replaced by string using `pandas.DataFrame.fillna()`  or `pandas.PivotTable(fill_value=)`

#### Expected Output

Expected that highlight will still work on column when `fillna` or `fill_value` get strings. 

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : None
python           : 3.7.3.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.0.0-27-generic
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 0.25.1
numpy            : 1.17.1
pytz             : 2019.2
dateutil         : 2.8.0
pip              : 19.2.3
setuptools       : 41.2.0
Cython           : None
pytest           : 5.1.1
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : 4.4.1
html5lib         : 1.0.1
pymysql          : None
psycopg2         : None
jinja2           : 2.10.1
IPython          : 7.7.0
pandas_datareader: None
bs4              : 4.8.0
bottleneck       : None
fastparquet      : None
gcsfs            : None
lxml.etree       : 4.4.1
matplotlib       : 3.1.1
numexpr          : None
odfpy            : None
openpyxl         : 2.6.3
pandas_gbq       : None
pyarrow          : None
pytables         : None
s3fs             : None
scipy            : 1.3.1
sqlalchemy       : 1.3.7
tables           : None
xarray           : None
xlrd             : 1.2.0
xlwt             : 1.3.0
xlsxwriter       : None

</details>

### PR DESCRIPTION
**Pull request by immaxchen**
As described in the issues, user who wants to control how NA values are printed
while applying styles to the output will have to implement their own formatter.
(so that the underlying data will not change and can be used for styling)

Since the behavior is common in styling (for reports etc.), suggest to add this
shortcut function to enable users format their NA values as something like '--'
or 'Not Available' easily.

EDIT:

- Change implementation to integrate with the original `.format()` using `na_rep` argument
- Add a new table-wise default `na_rep` setting, which can be set through `.set_na_rep()`
- Minor fix: formatter should be wrapped outside of locs-loop 
- Added a few user guide examples and test cases

example usage: `df.style.highlight_max().format(None, na_rep="-")`

- [x] closes #21527 and #28358
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: doc/source/whatsnew/v1.0.0.rst [hunk 1]
```diff
@@ -110,6 +110,7 @@ Other enhancements
 - :meth:`DataFrame.to_json` now accepts an ``indent`` integer argument to enable pretty printing of JSON output (:issue:`12004`)
 - :meth:`read_stata` can read Stata 119 dta files. (:issue:`28250`)
 - Added ``encoding`` argument to :func:`DataFrame.to_html` for non-ascii text (:issue:`28663`)
+- :meth:`Styler.format_null` is now added into the built-in functions to help formatting missing values (:issue:`28358`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.0.0.rst]: can u add this into the user guide as well

Comment (topper-123) [reply doc/source/whatsnew/v1.0.0.rst]: I'd like the name ``format_nans`` better, to be similar to ``fillna``, ``hasnans`` etc.

@jreback?

Comment (immaxchen) [reply doc/source/whatsnew/v1.0.0.rst]: Sure, I'll be glad to!

**Comment by jbrockmendel** [comment]
@topper-123 My understanding is that the "Style" tag is refers to code style, linting stuff, not the Styler class.  Do I have this backwards?

**Comment by immaxchen** [comment]
The doc has been added, please have a look! :smile:
How about the method name? do we have a conclusion?

File: pandas/io/formats/style.py [hunk 2]
```diff
@@ -930,6 +930,25 @@ def hide_columns(self, subset):
...
+
+        Parameters
+        ----------
+        na_rep : str
+
+        Returns
+        -------
+        self : Styler
+        """
+        self.format(
```
Comment (TomAugspurger) [inline pandas/io/formats/style.py]: This looks like it will overwrite the formatting of a previously applied formatter for non-NA values. Something like

```python
df.style.format("hi-{}".format).format_null()
```

is that the case?

**Review by TomAugspurger — COMMENTED** [review]
We've been moving away from null and to NA. So I think this should be called `format_na` if anything.

I'm a bit concerned about the implementation. I suspect this would be better handled as an argument to `Styler` and a method like `set_na_format` (see how uuid is handled)`. Then the `default_display_func` would check if the value is NA and use the na format if it's NA.

File: pandas/io/formats/style.py (hunk 2 above)
Comment (immaxchen) [reply pandas/io/formats/style.py]: Thanks @TomAugspurger, I like the name `format_na`! and yes, I was intended to make an overwriting implementation. Actually, I have considered the `set_*` approach as you, but it seems confusing for the case:
```
df.style.format('{:.2%}').set_na_format('-') # got 'nan%' instead of '-'
```
I've got a new idea, how about interface like this?
```
.format_na('-', subset=['col1','col2'])
.format('{:.2%}', na_rep='-', subset=['col3','col4'])
```
And the docstring for `format_na` rephrase to:
```
Format the text display value using default formatter but represent nan as `na_rep`.
For more advanced formatting, use Styler.format() with your custom formatter.
```

Comment (TomAugspurger) [reply pandas/io/formats/style.py]: I may not have been clear about my concern. It's fine that `na_format` overwrites the formatting for NA values. I'm concerned tht it overwrites the formatting for *non*-NA values. In my `.format("hi-{}".format).format_na('NA')` example, the NA values should be formatted as 'NA' and the non-NA values should be formatted as `hi-<value>`. But I suspect that right now the non-NA formatting is lost (though perhaps it's not).

Comment (TomAugspurger) [reply pandas/io/formats/style.py]: Not sure about adding an `na_rep` to the `.format` function... That's probably fine. I think it'd still be useful for users to have a way to control the default NA formatting at the table level.

But if we add an `na_rep` to `format`, then we wouldn't need a new `format_na` method, right?

Comment (immaxchen) [reply pandas/io/formats/style.py]: Sounds good, so a setting at the table level: `Styler.na_rep`
and the `def format(self, formatter, subset=None):` becomes 
`def format(self, formatter=None, na_rep=None, subset=None):`
drop `.format_na('-')`, use `.format(na_rep='-')` instead, right?

Comment (TomAugspurger) [reply pandas/io/formats/style.py]: I think that sounds correct. I'm not sure what the default should be, but probably just None (no special formatting for NA values).

**Comment by immaxchen** [comment]
Revision done, please help to review, thanks!  :D

**Comment by immaxchen** [comment]
I think this PR can also resolve #21527 (ENH: Styler display of NaN / null Values)

**Comment by TomAugspurger** [comment]
Most likely not for a couple weeks.

**Comment by immaxchen** [comment]
Ouch. :joy: 
@jreback can you help to review? :grin:

File: pandas/io/formats/style.py [hunk 3]
```diff
@@ -416,16 +423,20 @@ def format_attr(pair):
             table_attributes=table_attr,
         )
 
-    def format(self, formatter, subset=None):
+    def format(self, formatter=None, subset=None, na_rep=None):
```
Comment (WillAyd) [inline pandas/io/formats/style.py]: What was the point of changing `formatter` here?

File: pandas/io/formats/style.py [hunk 4]
```diff
@@ -891,6 +908,23 @@ def set_table_styles(self, table_styles):
         self.table_styles = table_styles
         return self
 
+    def set_na_rep(self, na_rep):
```
Comment (WillAyd) [inline pandas/io/formats/style.py]: Can you annotate this function? I think should be

```suggestion
    def set_na_rep(self, na_rep: str) -> "Styler":
```

File: pandas/io/formats/style.py [hunk 5]
```diff
@@ -935,19 +969,21 @@ def _highlight_null(v, null_color):
             "background-color: {color}".format(color=null_color) if pd.isna(v) else ""
         )
 
-    def highlight_null(self, null_color="red"):
+    def highlight_null(self, null_color="red", subset=None):
```
Comment (WillAyd) [inline pandas/io/formats/style.py]: How is this related to `na_rep`?

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/io/formats/style.py (hunk 5 above)
Comment (immaxchen) [reply pandas/io/formats/style.py]: Actually not related. :joy: 
Just making it consistent with `highlight_min` and `highlight_max`, also for completeness of missing values styling and formatting.

Comment (WillAyd) [reply pandas/io/formats/style.py]: Nice idea but would prefer if you take this out then. We try to keep PRs to changing one thing at a time as it helps speed up review process

Comment (WillAyd) [reply pandas/io/formats/style.py]: Can always do as a follow up PR

Comment (immaxchen) [reply pandas/io/formats/style.py]: Thanks, that is a great advice!

**Comment by immaxchen** [comment]
Revision done, please have a look. :grin:

File: pandas/io/formats/style.py [hunk 6]
```diff
@@ -71,6 +71,11 @@ class Styler:
         The ``id`` takes the form ``T_<uuid>_row<num_row>_col<num_col>``
         where ``<uuid>`` is the unique identifier, ``<num_row>`` is the row
         number and ``<num_col>`` is the column number.
+    na_rep : str or None, default None
```
Comment (TomAugspurger) [inline pandas/io/formats/style.py]: I think this should be `str, optional`.

File: pandas/io/formats/style.py [hunk 7]
```diff
@@ -1480,3 +1518,13 @@ def _maybe_wrap_formatter(formatter):
             "instead".format(formatter=formatter)
         )
         raise TypeError(msg)
+
+
+def _maybe_wrap_na_formatter(formatter, na_rep):
```
Comment (TomAugspurger) [inline pandas/io/formats/style.py]: Can this be folded into `_maybe_wrap_formatter`? I'm finding the multiple layers of wrapping hard to follow.

**Review by TomAugspurger — COMMENTED** [review]
Thanks. Can you add

1. A test exercising passing na_rep to Styler, `Styler(df, na_rep="-")`
2. A test ensuring that non-numeric NA values are handled correctly? At least NaT for datetime data.

File: pandas/io/formats/style.py (hunk 7 above)
Comment (immaxchen) [reply pandas/io/formats/style.py]: No problem, all done!

**Comment by immaxchen** [comment]
Hi Tom, William, please have a look at the revision when you are available, thanks~

File: doc/source/whatsnew/v1.0.0.rst [hunk 8]
```diff
@@ -114,6 +114,8 @@ Other enhancements
 - Added ``encoding`` argument to :meth:`DataFrame.to_string` for non-ascii text (:issue:`28766`)
 - Added ``encoding`` argument to :func:`DataFrame.to_html` for non-ascii text (:issue:`28663`)
 - :meth:`Styler.background_gradient` now accepts ``vmin`` and ``vmax`` arguments (:issue:`12145`)
+- :class:`Styler` added :meth:`Styler.set_na_rep` method to set default missing values representation for the entire table.
+  :meth:`Styler.format` added the ``na_rep`` parameter to help format the missing values (:issue:`21527`, :issue:`28358`)
```
Comment (WillAyd) [inline doc/source/whatsnew/v1.0.0.rst]: I think just this note is fine (can delete the line above)

File: pandas/io/formats/style.py [hunk 9]
```diff
@@ -126,6 +131,7 @@ def __init__(
         caption=None,
         table_attributes=None,
         cell_ids=True,
+        na_rep=None,
```
Comment (WillAyd) [inline pandas/io/formats/style.py]: Can you annotate this? I think Optional[str]

File: pandas/io/formats/style.py [hunk 10]
```diff
@@ -416,16 +425,22 @@ def format_attr(pair):
             table_attributes=table_attr,
         )
 
-    def format(self, formatter, subset=None):
+    def format(self, formatter, subset=None, na_rep=None):
```
Comment (WillAyd) [inline pandas/io/formats/style.py]: annotation here as well

File: pandas/io/formats/style.py [hunk 11]
```diff
@@ -1487,14 +1523,22 @@ def _get_level_lengths(index, hidden_elements=None):
     return non_zero_lengths
 
 
-def _maybe_wrap_formatter(formatter):
+def _maybe_wrap_formatter(formatter, na_rep):
```
Comment (WillAyd) [inline pandas/io/formats/style.py]: and here

File: pandas/io/formats/style.py [hunk 12]
```diff
@@ -1487,14 +1523,22 @@ def _get_level_lengths(index, hidden_elements=None):
...
             "instead".format(formatter=formatter)
         )
         raise TypeError(msg)
+
+    if na_rep is None:
+        return formatter_func
+    elif isinstance(na_rep, str):
+        return lambda x: na_rep if pd.isna(x) else formatter_func(x)
+    else:
+        msg = "Expected a string, got {na_rep} instead".format(na_rep=na_rep)
```
Comment (WillAyd) [inline pandas/io/formats/style.py]: Can you add a test case that hits this?

**Review by WillAyd — CHANGES_REQUESTED** [review]

**Comment by immaxchen** [comment]
After adding the annotation for `na_rep`, mypy complained as below, so I have no choice but to add more annotations.
```
pandas/io/formats/style.py:136: error: Need type annotation for 'ctx'
pandas/io/formats/style.py:137: error: Need type annotation for '_todo' (hint: "_todo: List[<type>] = ...")
pandas/io/formats/style.py:158: error: Need type annotation for 'hidden_columns' (hint: "hidden_columns: List[<type>] = ...")
pandas/io/formats/style.py:173: error: Need type annotation for '_display_funcs'
pandas/io/formats/style.py:470: error: "None" not callable
```

File: pandas/io/formats/style.py [hunk 13]
```diff
@@ -126,9 +131,10 @@ def __init__(
         caption=None,
         table_attributes=None,
         cell_ids=True,
+        na_rep: Optional[str] = None,
     ):
-        self.ctx = defaultdict(list)
-        self._todo = []
+        self.ctx = defaultdict(list)  # type: DefaultDict[Tuple[int, int], List[str]]
+        self._todo = []  # type: List[Tuple[Callable, Tuple, Dict]]
```
Comment (WillAyd) [inline pandas/io/formats/style.py]: Is it possible to add sub-types for Callable, Tuple, Dict here?

File: pandas/io/formats/style.py [hunk 14]
```diff
@@ -416,16 +427,22 @@ def format_attr(pair):
             table_attributes=table_attr,
         )
 
-    def format(self, formatter, subset=None):
+    def format(self, formatter, subset=None, na_rep: Optional[str] = None):
```
Comment (WillAyd) [inline pandas/io/formats/style.py]: Not required here but if you wanted to in a follow up annotate other arguments would be super helpful

File: pandas/tests/io/formats/test_style.py [hunk 15]
```diff
@@ -1009,6 +1009,75 @@ def test_bar_bad_align_raises(self):
...
+        assert ctx["body"][0][1]["display_value"] == "-"
+        assert ctx["body"][0][2]["display_value"] == "-"
+        assert ctx["body"][1][1]["display_value"] == "-"
+        assert ctx["body"][1][2]["display_value"] == "-"
+
+    @pytest.mark.xfail
+    def test_format_with_bad_na_rep(self):
+        # GH 21527 28358
+        df = pd.DataFrame([[None, None], [1.1, 1.2]], columns=["A", "B"])
+        df.style.format(None, na_rep=-1)
```
Comment (WillAyd) [inline pandas/tests/io/formats/test_style.py]: Instead of marking this as fail should use pytest.raises as a context manager (you'll find other examples throughout tests)

**Review by WillAyd — CHANGES_REQUESTED** [review]
Looks good. Minor issue with the new test rest of the comments are optional and can be done as follow ups

**Comment by immaxchen** [comment]
Thanks @WillAyd, I have revised using `pytest.raises` and resolve merge conflict. Since this PR is getting quite lengthy, I will try to add annotations in the follow-ups.

**Review by WillAyd — CHANGES_REQUESTED** [review]
OK last edit - I know you want to refine types further later but in the interim can you convert to the Py36 syntax? We have #29741 outstanding which is doing that for existing types, so want to use that in new PRs as well

**Comment by immaxchen** [comment]
Hi @WillAyd, revision done for py36 type syntax, please have a look :)

**Review by WillAyd — APPROVED** [review]
Nice PR this lgtm. @TomAugspurger care to look?

**Review by TomAugspurger — APPROVED** [review]

**Comment by TomAugspurger** [comment]
Thanks @immaxchen!
